src/frequenz/sdk/util/ringbuffer.py

```diff
@@ -0,0 +1,412 @@
+        return self._missing_windows
+
+    @property
+    def maxlen(self) -> int:
+        """Get the max length.
+
+        Returns:
+            the max amount of items this container can hold.
+        """
+        return len(self._buffer)
+
+    def update(self, timestamp: datetime, value: T, missing: bool = False) -> bool:
+        """Update the buffer with a new value for the given timestamp.
+
+        Args:
+            timestamp: Timestamp of the new value.
+            value: value to add.
+            missing: if true, the given timestamp will be recorded as missing.
+                The value will still be written.
+
+                Note: When relying on your own values to mark missing data (e.g.
+                `math.nan`), be advised that when `update()` is called with
+                timestamps newer than the current time + `sampling_period_s`
+                (as is the case for loading historical data after an app
+                restart), a gap of missing data exists.
+                This gap does not contain NaN values but simply the old invalid values.
+                The `missing_windows` will reflect this.
+
+        Returns:
+            True if the entry was added. False if it was discarded for being too old.
+        """
+        # adjust timestamp to be exactly on the sample period time point
+        timestamp = self.normalize_timestamp(timestamp)
+
+        # Don't add outdated entries
+        if timestamp < self._datetime_oldest and self._datetime_oldest != datetime.max:
+            return False
+
+        # print(f"Normalizing {timestamp} to {self.normalize_timestamp(timestamp)}")
+
+        # Update timestamps
+        prev_newest = self._datetime_newest
+        self._datetime_newest = max(self._datetime_newest, timestamp)
+        self._datetime_oldest = self._datetime_newest - self._time_range
+
+        # Update data
+        self._buffer[self.datetime_to_index(timestamp)] = value
+
+        self._update_missing_windows(timestamp, prev_newest, missing)
+
+        return True
+
+    def datetime_to_index(
+        self, timestamp: datetime, allow_outside_range: bool = False
+    ) -> int:
+        """Convert the given timestamp to an index.
+
+        Args:
+            timestamp: Timestamp to convert.
+            allow_outside_range: If True, don't throw an exception when the
+                timestamp is outside our bounds
+
+        Raises:
+            IndexError: when requesting a timestamp outside the range this container holds
+
+        Returns:
+            index where the value for the given timestamp can be found.
+        """
+        timestamp = self.normalize_timestamp(timestamp)
+
+        if not allow_outside_range and (
+            self._datetime_newest + timedelta(seconds=1 * self._sampling_period_s)
+            < timestamp
+            or timestamp < self._datetime_oldest
+        ):
+            raise IndexError(
+                f"Requested timestamp {timestamp} is is "
```

Comment:
`Requested timestamp {timestamp} is is ` - > `Requested timestamp {timestamp} is `